crates/multithread/src/multithread.rs

```diff
@@ -0,0 +1,270 @@
+use e3_trbfv::calculate_decryption_share::calculate_decryption_share;
+use e3_trbfv::calculate_threshold_decryption::calculate_threshold_decryption;
+use e3_trbfv::gen_esi_sss::gen_esi_sss;
+use e3_trbfv::gen_pk_share_and_sk_sss::gen_pk_share_and_sk_sss;
+use e3_trbfv::{TrBFVError, TrBFVRequest, TrBFVResponse};
+use e3_utils::SharedRng;
+use rand::Rng;
+use tracing::error;
+use tracing::info;
+
+/// Multithread actor
+pub struct Multithread {
+    bus: BusHandle,
+    rng: SharedRng,
+    cipher: Arc<Cipher>,
+    task_pool: TaskPool,
+    report: Option<Addr<MultithreadReport>>,
+}
+
+impl Multithread {
+    pub fn new(
+        bus: BusHandle,
+        rng: SharedRng,
+        cipher: Arc<Cipher>,
+        task_pool: TaskPool,
+        report: Option<Addr<MultithreadReport>>,
+    ) -> Self {
+        Self {
+            bus,
+            rng,
+            cipher,
+            task_pool,
+            report,
+        }
+    }
+
+    /// Subtract the given amount from the total number of available threads and return the result
+    pub fn get_max_threads_minus(amount: usize) -> usize {
+        let total_threads = thread::available_parallelism()
+            .map(|n| n.get())
+            .unwrap_or(1);
+        let threads_to_use = std::cmp::max(1, total_threads.saturating_sub(amount));
+        threads_to_use
+    }
+
+    pub fn attach(
+        bus: &BusHandle,
+        rng: SharedRng,
+        cipher: Arc<Cipher>,
+        task_pool: TaskPool,
+        report: Option<Addr<MultithreadReport>>,
+    ) -> Addr<Self> {
+        let addr = Self::new(bus.clone(), rng.clone(), cipher.clone(), task_pool, report).start();
+        bus.subscribe("ComputeRequest", addr.clone().recipient());
+        addr
+    }
+
+    pub fn create_taskpool(threads: usize, max_tasks: usize) -> TaskPool {
+        TaskPool::new(threads, max_tasks)
+    }
+}
+
+impl Actor for Multithread {
+    type Context = actix::Context<Self>;
+}
+
+impl Handler<EnclaveEvent> for Multithread {
+    type Result = ();
+    fn handle(&mut self, msg: EnclaveEvent, ctx: &mut Self::Context) -> Self::Result {
+        info!("Multithread received EnclaveEvent!");
+        match msg.get_data() {
+            EnclaveEventData::ComputeRequest(data) => ctx.notify(data.clone()),
+            _ => (),
+        }
+    }
+}
+
+impl Handler<ComputeRequest> for Multithread {
+    type Result = ResponseFuture<()>;
+    fn handle(&mut self, msg: ComputeRequest, _: &mut Self::Context) -> Self::Result {
+        let cipher = self.cipher.clone();
+        let rng = self.rng.clone();
+        let bus = self.bus.clone();
+        let pool = self.task_pool.clone();
+        let report = self.report.clone();
+        // TODO: replace with trap_fut
+        Box::pin(async move {
+            match handle_compute_request_event(msg, bus, cipher, rng, pool, report).await {
+                Ok(_) => (),
+                Err(e) => error!("{e}"),
+            }
+        })
```

Comment:
If the request panics or the semaphore is closed, then wouldnt this hang forever? cuz only `error!()` is called and `ThresholdKeyshare` stays in `GeneratingThresholdShare` state forever. 
We should publish the error to the bus here no? unless im missing something and it does it elsewhere